crates/tests/test-tedge-cli/src/lib.rs

```diff
@@ -0,0 +1,96 @@
+mod test_env;
+
+#[cfg(test)]
+mod tests {
+    use crate::*;
+    use assert_matches::*;
+    use rexpect::errors::*;
+    use rexpect::process::signal::Signal;
+    use rexpect::process::wait::WaitStatus;
+    use rexpect::*;
+
+    macro_rules! execute_cmd {
+        ($($arg:tt)*) => {spawn(&format!($($arg)*), TIMEOUT_MS)?.process.wait()?}
+    }
+
+    macro_rules! spawn_cmd {
+        ($($arg:tt)*) => {spawn(&format!($($arg)*), TIMEOUT_MS)?}
+    }
+
+    const TIMEOUT_MS: Option<u64> = Some(5_000);
+
+    #[test]
+    fn it_works() -> Result<()> {
+        let mut sub = spawn("tedge mqtt sub test/topic", TIMEOUT_MS)?;
+
+        execute(r#"tedge mqtt pub test/topic "hello thin-edge""#)?;
+        assert_eq!(sub.read_line()?, "INFO: Connected");
+        assert_eq!(sub.read_line()?, "[test/topic] hello thin-edge");
+
+        execute(r#"tedge mqtt pub test/topic bye-bye --qos 2"#)?;
+        assert_eq!(sub.read_line()?, "[test/topic] bye-bye");
+
+        sub.process.kill(Signal::SIGTERM)?;
+        Ok(())
+    }
+
+    #[test]
+    fn tedge_cert_upload_request_a_password() -> Result<()> {
+        // Create a certificate
+        execute("tedge config set device.cert.path /tmp/test-device.cert")?;
+        execute("tedge config set device.key.path /tmp/test-device.key")?;
+        execute("tedge cert create --device-id test-device")?;
+
+        // Make sure the certificate has been created
+        let mut sub = spawn("tedge config get device.id", TIMEOUT_MS)?;
+        assert_eq!(sub.read_line()?, "test-device");
+        sub.process.wait()?;
+
+        // Upload the certificate
+        execute("tedge config set c8y.url didier.latest.stage.c8y.io")?;
+        let mut sub = spawn("tedge cert upload c8y --user foo-user", TIMEOUT_MS)?;
+        sub.exp_string("Enter password: ")?;
+        sub.send_line("foo-password")?;
+        assert_eq!(sub.read_line()?, "");
```

Comment:
Here some magic from `rexpect`: we don't need a trick for the test and the password is passed via a `tty`.